lib/Monads/nondet/Nondet_VCG.thy

```diff
@@ -553,24 +566,48 @@ lemma hoare_vcg_disj_lift:
   unfolding valid_def
   by fastforce
 
+lemma hoare_vcg_disj_lift_R:
+  assumes x: "\<lbrace>P\<rbrace> f \<lbrace>Q\<rbrace>,-"
+  assumes y: "\<lbrace>P'\<rbrace> f \<lbrace>Q'\<rbrace>,-"
+  shows      "\<lbrace>\<lambda>s. P s \<or> P' s\<rbrace> f \<lbrace>\<lambda>rv s. Q rv s \<or> Q' rv s\<rbrace>,-"
+  using assms
+  by (fastforce simp: validE_R_def validE_def valid_def split: sum.splits)
+
 lemma hoare_vcg_const_Ball_lift:
   "\<lbrakk> \<And>x. x \<in> S \<Longrightarrow> \<lbrace>P x\<rbrace> f \<lbrace>Q x\<rbrace> \<rbrakk> \<Longrightarrow> \<lbrace>\<lambda>s. \<forall>x\<in>S. P x s\<rbrace> f \<lbrace>\<lambda>rv s. \<forall>x\<in>S. Q x rv s\<rbrace>"
   by (fastforce simp: valid_def)
 
+lemma hoare_vcg_const_Ball_liftE:
+  "\<lbrakk> \<And>x. x \<in> S \<Longrightarrow> \<lbrace>P x\<rbrace> f \<lbrace>Q x\<rbrace>,\<lbrace>E\<rbrace>; \<lbrace>\<lambda>s. True\<rbrace> f \<lbrace>\<lambda>r s. True\<rbrace>, \<lbrace>E\<rbrace> \<rbrakk> \<Longrightarrow> \<lbrace>\<lambda>s. \<forall>x\<in>S. P x s\<rbrace> f \<lbrace>\<lambda>rv s. \<forall>x\<in>S. Q x rv s\<rbrace>,\<lbrace>E\<rbrace>"
+  by (fastforce simp: validE_def valid_def split: sum.splits)
+
 lemma hoare_vcg_const_Ball_lift_R:
  "\<lbrakk> \<And>x. x \<in> S \<Longrightarrow> \<lbrace>P x\<rbrace> f \<lbrace>Q x\<rbrace>,- \<rbrakk> \<Longrightarrow> \<lbrace>\<lambda>s. \<forall>x \<in> S. P x s\<rbrace> f \<lbrace>\<lambda>rv s. \<forall>x \<in> S. Q x rv s\<rbrace>,-"
   unfolding validE_R_def validE_def
   by (rule hoare_strengthen_post)
      (fastforce intro!: hoare_vcg_const_Ball_lift split: sum.splits)+
 
+lemma hoare_vcg_const_Ball_lift_E_E:
```

Comment:
Just confirming the naming between `fooE`, `foo_R`, and `foo_E_E`: is this the way we want it? (no underscore for the `E` wp variant and with underscore for the specialised forms)

Reply:
Good spotting, I missed that the first time through. I have some more improvements on the way that includes these lemmas; I have been trying to standardise on `foo`, `fooE`, `fooE_R`, and `fooE_E`.

Reply:
Cool, happy for that to go in a  separate PR.